tools/cldr-code/src/main/java/org/unicode/cldr/test/CheckDates.java

```diff
@@ -654,6 +662,70 @@ && isTooMuchWiderThan(value, abbrValue)
         return this;
     }
 
+    private void checkTimeFormatMatchesRegion(String value, List<CheckStatus> result) {
+        DateTimePatternGenerator dtpg = DateTimePatternGenerator.getEmptyInstance();
+        Map<String /* region */, PreferredAndAllowedHour> timeData = sdi.getTimeData();
+        Map<String, String> likelySubtags = sdi.getLikelySubtags();
+        String localeID = getResolvedCldrFileToCheck().getLocaleID();
+        String jPattern = getRegionHourFormat(timeData, localeID, likelySubtags);
+        if (jPattern == null) {
+            CheckStatus item =
+                    new CheckStatus()
+                            .setCause(this)
+                            .setMainType(CheckStatus.errorType)
+                            .setSubtype(Subtype.inconsistentTimePattern)
+                            .setMessage("No hour format found");
+            result.add(item);
+            return;
+        }
+        String shortPatSkeleton = dtpg.getBaseSkeleton(value); // e.g., "ahm" or "Hm"
+        String jPatSkeleton = dtpg.getBaseSkeleton(jPattern); // e.g., "ah" or "H"
+        final char[] timeCycleChars = {'H', 'h', 'K', 'k'};
+        for (char timeCycleChar : timeCycleChars) {
+            if (jPatSkeleton.indexOf(timeCycleChar) >= 0
+                    && shortPatSkeleton.indexOf(timeCycleChar) < 0) {
+                String message =
+                        "Time format does not match region; expected "
+                                + timeCycleChar
+                                + " in the value "
+                                + value;
+                CheckStatus item =
+                        new CheckStatus()
+                                .setCause(this)
+                                .setMainType(CheckStatus.errorType)
+                                .setSubtype(Subtype.inconsistentTimePattern)
+                                .setMessage(message);
+                result.add(item);
+                return;
+            }
+        }
+    }
+
+    private String getRegionHourFormat(
+            Map<String, PreferredAndAllowedHour> timeData,
+            String localeID,
+            Map<String, String> likelySubtags) {
+        PreferredAndAllowedHour prefAndAllowedHr = timeData.get(localeID);
+        if (prefAndAllowedHr == null) {
+            LocaleIDParser lp = new LocaleIDParser();
+            String region = lp.set(localeID).getRegion();
+            if (region == null || region.isEmpty()) {
+                String loc2 = likelySubtags.get(localeID);
+                if (loc2 != null && !loc2.isEmpty()) {
+                    region = lp.set(loc2).getRegion();
+                }
```

Comment:
We need one more step here. At this point we know localeID does not have a region, and we tried to get the likelySubtags for it in loc2. However, if that fails, it might be because localeID has a script, and likelySubtags does not have an entry for that combination of language and script. In that case we have to strip the script and look up likelySubtags for the language part only, and then get the region from that.

Reply:
Thanks! I'll follow up with this soon